File: probs/dmoj/bfs17/p6.py
# use 2d bit or smth
n = int(input())

a = []
for i in range(n):
    l, w, h, p = map(int, input().split())
    if l < w:
        k = w
        w = l 
        l = k 
    a.append((h,l,w,p,i+1))

bit = [[(0, 0) for i in range(5005)] for i in range(5005)]

def updy(x, y, val):
    while y <= 5000:
        if val[0] > bit[x][y][0]:
            bit[x][y] = val
        y += (y&-y)

def upd(x, y, val):
    while x <= 5000:
        updy(x, y, val)
        x += (x&-x)

def quey(x, y):
    ret = 0
    val = -1
    while y > 0:
        if bit[x][y][0] > ret:
            ret = bit[x][y][0]
            val = bit[x][y][1]
        y -= (y&-y)
    return (ret,val)

def que(x, y):
    ret = 0
    val = -1
    while x > 0:
        k = quey(x,y)
        if k[0] > ret:
            ret = k[0]
            val = k[1]
        x -= (x&-x)
    return (ret,val)

a.sort()
last = [-1 for i in range(n)]
ans = [0 for i in range(n)]
for i in range(n):
    p = que(a[i][1], a[i][2])
    last[i] = p[1]
    tot = a[i][3]+p[0]
    ans[i] = tot
    upd(a[i][1], a[i][2], (tot, i))

mxi = -1
bes = 0
for i in range(n):
    if ans[i] > bes:
        mxi = i
        bes = ans[i]

print(bes)

l = []
while mxi != -1:
    l.append(a[mxi][4])
    mxi = last[mxi]

print(len(l))
print(*l, sep=' ')

File: probs/dmoj/bfs17/p6.cpp
#include <bits/stdc++.h>
using namespace std;
using pii = pair<int,int>;
using pli = pair<long long, int>;
#define f first
#define s second

struct B 
{
    signed l, w, h, p, i;
    B() {}
    B(int l, int w, int h, int p, int i)
    {
        this->l = l;
        this->w = w;
        this->h = h;
        this->p = p;
        this->i = i;
    }
    bool operator<(B other)
    {
        if(h == other.h)
        {
            if(l == other.l) return w < other.w;
            return l < other.l;
        }
        return h < other.h;
    }
};

int n, last[100005], ans[100005];
B a[100005];
pli bit[5005][5005];

void updy(int x, int y, pii val)
{
	while(y <= 1000)
	{
		if(bit[x][y].f < val.f)
        {
            bit[x][y] = val;
        }
		y += (y&-y);
	}
}
 
void upd(int x, int y, pii val)
{
	while(x <= 1000)
	{
		updy(x, y, val);
		x += (x&-x);
	}
}
 
pli quey(int x, int y)
{
	long long ret = 0;
    int val = 0;
	while(y > 0)
	{
        if(ret < bit[x][y].f)
        {
		    ret = bit[x][y].f;
            val = bit[x][y].s;
        }
		y -= (y & -y);
	}
	return {ret,val};
}
 
pli que(int x, int y)
{
	long long ret = 0;
    int val = 0;
	while(x > 0)
	{
		pii k = quey(x, y);
        if(ret < k.f)
        {
            ret = k.f;
            val = k.s;
        }
		x -= (x & -x);
	}
	return {ret,val};
}

signed main()
{
    ios_base::sync_with_stdio(0);
    cin.tie(NULL);
    cout.tie(NULL);
    cin >> n;
    for(int i = 0; i < n; i++)
    {
        int l, w, h, p; cin >> l >> w >> h >> p;
        if(l<w) swap(l,w);
        a[i] = B(l, w, h, p, i+1);
    }
    sort(a, a+n);
    fill(last, last+n, 0);
    for(int i = 0; i < n; i++)
    {
        pii p = que(a[i].l, a[i].w);
        last[i] = p.s;
        ans[i] = a[i].p+p.f;
        upd(a[i].l, a[i].w, {ans[i],i+1});
    }

    int mxi = 0;
    int bes = 0;
    for(int i = 0; i < n; i++)
    {
        if(ans[i] > bes)
        {
            mxi = i;
            bes = ans[i];
        }
    }
    cout << bes << endl;

    vector<signed> l;
    while(mxi != -1)
    {
        l.push_back(a[mxi].i);
        mxi = last[mxi]-1;
    }
    cout << l.size() << endl;
    for(auto& x : l) cout << x << " ";
    cout << endl;
}